Support search traces by prompt version

TomeHirata opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/TomeHirata/mlflow/pull/18906?quickstart=1)

#### Install mlflow from this PR

```
# mlflow
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/18906/merge
# mlflow-skinny
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/18906/merge#subdirectory=libs/skinny
```

For Databricks, use the following command:

```
%sh curl -LsSf https://raw.githubusercontent.com/mlflow/mlflow/HEAD/dev/install-skinny.sh | sh -s pull/18906/merge
```

</p>
</details>

### Related Issues/PRs

n/a

### What changes are proposed in this pull request?

This PR introduces a new `prompts` filter for `search_traces`. As part is this change, we migrate the linkage storage from trace tag to entity association table of tracking store.

```
traces = mlflow.search_traces(filter_string="prompts = 'prompt-a/1'")
```

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [x] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [x] Instructions

### Release Notes

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

<!-- Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change. -->

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/prompts`: MLflow prompt engineering features, prompt templates, and prompt management
- [x] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

<!-- Do not modify or remove any text inside the parentheses. Keep both checkboxes below. -->

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

TomeHirata committed:
Refactor linked prompts handling to use TraceTagKey

- Removed the deprecated LINKED_PROMPTS_TAG_KEY from prompt constants.
- Updated all references to linked prompts in the codebase to use TraceTagKey.LINKED_PROMPTS.
- Added validation for prompts filter in search functionality to ensure only exact matches are allowed.
- Enhanced error handling for invalid prompts filter format.

This change centralizes the management of linked prompts and improves the clarity of the codebase.

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

TomeHirata committed:
Enhance search_traces documentation to include linked prompts filter

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

TomeHirata committed:
Merge branch 'master' into feat/prompt/search-traces

B-Step62 commented on docs/docs/genai/tracing/search-traces.mdx:
```diff
@@ -55,19 +55,20 @@ The `search_traces` API uses a SQL-like Domain Specific Language (DSL) for query
+| **Trace Timestamps** | `trace.timestamp_ms`, `trace.execution_time_ms`, `trace.end_time_ms` | `=`, `!=`, `>`, `<`, `>=`, `<=`     | trace.end_time_ms > 1762408895531         |
+| **Trace IDs**        | `trace.run_id`                                                       | `=`                                 | trace.run_id = "run_id"                   |
+| **String Fields**    | `trace.client_request_id`, `trace.name`                              | `=`, `!=`, `LIKE`, `ILIKE`, `RLIKE` | trace.name LIKE "%Generate%"              |
+| **Linked Prompts**   | `prompts`                                                            | `=` (format: `"name/version"`)      | prompts = "qa-system-prompt/4"            |
```
```suggestion
| **Linked Prompts**   | `prompt`                                                            | `=` (format: `"name/version"`)      | prompt = "qa-system-prompt/4"            |
```

Is plural intentional? Since we only match with a single prompt version, singular sounds more natural.

B-Step62 commented on docs/docs/genai/tracing/search-traces.mdx:
Can we add some documentation under `/genai/prompt-registry/` as well? Maybe it's worth having a single page explains link between prompt and traces. Maybe follow-up after adding UI feature to jump to a list of traces from the prompt page.

TomeHirata replied:
Yeah, let me add the doc section after adding UI changes.

TomeHirata committed:
Update search_traces filter from 'prompts' to 'prompt' for consistency across documentation and codebase

- Changed all instances of 'prompts' to 'prompt' in the search_traces API documentation and related code.
- Updated error messages and test cases to reflect the new filter name.
- Ensured that the filter only supports exact matches with the format "name/version".

This change enhances clarity and consistency in the API usage.

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

TomeHirata committed:
Merge branch 'feat/prompt/search-traces' of github.com:TomeHirata/mlflow into feat/prompt/search-traces

TomeHirata committed:
Refactor prompts filter validation to ensure correct format

- Updated the filter validation logic to check for exactly one '/' in the prompt value, improving error handling for invalid inputs.
- Adjusted the test cases to include a new scenario for invalid prompt format.

This change enhances the robustness of the prompts filter functionality.

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

TomeHirata commented on mlflow/store/tracking/sqlalchemy_store.py:
```diff
@@ -4917,6 +4917,39 @@ def _get_filter_clauses_for_search_traces(filter_string, session, dialect):
+                        )
+                    parts = value.rsplit("/", 1)
+                    prompt_name = parts[0]
+                    prompt_version = parts[1]
```
Good call, updated the validation to cover that case

TomeHirata committed:
Add linkPromptsToTrace API and related functionality

- Introduced the `linkPromptsToTrace` RPC to associate multiple prompt versions with a trace, enhancing prompt tracking workflows.
- Implemented the necessary data structures and request handlers to support this functionality.
- Updated the tracking store interfaces to allow linking prompts to traces using entity associations.
- Added tests to verify the correct behavior of the new API and its integration with the existing system.

This change improves the ability to manage prompt versions in relation to traces, facilitating better tracking and organization of prompt usage.

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

TomeHirata committed:
Update linkPromptsToTrace API documentation for clarity

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

TomeHirata committed:
Implement backward compatibility for LINKED_PROMPTS tag in SqlAlchemyStore

- Added a method to populate the LINKED_PROMPTS tag from entity associations for traces that lack this tag, ensuring backward compatibility.
- Updated the get_trace_info method to call the new population method.
- Enhanced search_traces to include the LINKED_PROMPTS tag for relevant traces.
- Added unit tests to verify the correct population of LINKED_PROMPTS in various scenarios.

This change improves trace management by maintaining compatibility with existing data structures.

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

TomeHirata committed:
Enhance trace information population in SqlAlchemyStore

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

TomeHirata commented on mlflow/store/tracking/sqlalchemy_store.py:
```diff
@@ -4917,6 +4917,39 @@ def _get_filter_clauses_for_search_traces(filter_string, session, dialect):
 
             if SearchTraceUtils.is_tag(key_type, comparator):
                 entity = SqlTraceTag
+                # Special handling for prompts filter: only support exact match with name/version
```
I got your point, but do you think the single additional query for retrieving associations is that problematic? As you said earlier, it should be quick since it relies on database indices, and as I implemented, it does not cause N+1 issue by using batch loading. Not objecting option 2, we can go with dual writing if skipping one query is worth database denormalization.

B-Step62 replied:
I'm fine with populating it to tags if you feel strongly about it. For the TODO comment, I didn't expect to have "linked prompt" tab in the trace UI but just to have a button like "see linked prompts" that jumps to search page:)

TomeHirata committed:
Merge remote-tracking branch 'mlflow/master' into feat/prompt/search-traces

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

TomeHirata committed:
Add back tag based logging

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

TomeHirata committed:
fix test

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

TomeHirata committed:
lint

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

B-Step62 reviewed (approved):
LGTM!
